### ISSUE TITLE
stage1 AND stage2: C ABI issue returning C struct by value

### ISSUE THREAD

**Issue by mitchellh** [ziglang/zig#13125]
### Zig Version

0.10.0-dev.3998+c25ce5bba

On aarch64 (reproduced on both Mac and Linux)

### Steps to Reproduce

`test.c`:

```c
typedef struct Point {
    double x;
    double y;
} Point;

typedef struct Size {
    double width;
    double height;
} Size;

typedef struct Rect {
    Point origin;
    Size size;
} Rect;

Rect do_the_thing() {
    Rect x;
    x.origin.x = 1;
    x.origin.y = 2;
    x.size.width = 3;
    x.size.height = 4;
    return x;
}
```

`test.zig`:

```zig
const std = @import("std");

extern "c" fn do_the_thing() Rect;

// This is hand-written but matches identically what `translate-c` would do except
// I collapsed the structs inside. If you want to use `translate-c` you can, the bug
// still exists exactly the same.
const Rect = extern struct {
    origin: extern struct {
        x: f64,
        y: f64,
    },

    size: extern struct {
        width: f64,
        height: f64,
    },
};

test {
    std.log.warn("result={}", .{do_the_thing()});
}
```

Command:

`zig test -lc test.zig test.c` 

Also reproduces with `-fstage1`.

### Expected Behavior

I expect the log to output a Rect with x, y, width, height of 1, 2, 3, 4 respectively.

### Actual Behavior

Instead, it outputs some different garbage on every invocation, such as:

```
Test [1/1] test ""... [default] (warn): result=Rect{ .origin = struct:6:13{ .x = 1.390665929542936e-309, .y = 1.39066592953924e-309 }, .size = struct:11:11{ .width = 1.1047486e-317, .height = 1.39066592954098e-309 } }
All 1 tests passed.
```

**Comment by mitchellh** [comment]
Real world impact: this reproduction was made by mimicking an issue I saw using macOS system APIs. In this case, `Rect` is identical to `CGRect` (of Core Graphics) and `do_the_thing` is mimicking a function that returns the bounds of some graphical item. That's where I first noticed this behavior.

**Workaround:** I've done a workaround by writing a function in C that uses an out parameter and calling that from Zig. Basically:

```c
void fix_do_the_thing(Rect *out) { *out = do_the_thing(); }
```
